framework/include/actions/ActionWarehouse.h

```diff
@@ -101,6 +101,64 @@ class ActionWarehouse :
   const std::vector<Action *> & getActionsByName(const std::string & task) const;
 
   /**
+   * Retrieve an action with its name and the desired type.
+   * @param name The action name.
+   */
+  template <class T>
+  const T & getAction(const std::string & name)
+    {
+      T* p = NULL;
+      for (std::vector<std::string>::iterator it = _ordered_names.begin(); it != _ordered_names.end(); ++it)
+      {
+        std::string task = *it;
+        for (ActionIterator act_iter = actionBlocksWithActionBegin(task);
+             act_iter != actionBlocksWithActionEnd(task);
+             ++act_iter)
+        {
+          if ((*act_iter)->name() == name)
+          {
+            p = dynamic_cast<T*>(*act_iter);
+            if (!p)
+              mooseError("The action with name being "+name+" is not in the desired type");
+            else
+              break;
+          }
+        }
+      }
+      if (!p)
+        mooseError("Action with name being "+name+" does not exist");
+      return *p;
+    }
+
+  /**
+   * Retrieve all actions in a specific type.
+   */
+  template <class T>
+  std::vector<const T *> getActions()
```

Comment:
OK, this logic is better but I still have a problem with the decision on what to return. The method name is getActions and it returns a vector but you go through a lot of work to make sure that the vector contains unique pointers. 

I would prefer that you either just return the std::set if you really want unique pointers, or you name the method getUniqueActions() or you just return the vector with possible duplicates (of which I don't believe there should be any). Have you tried any of this out? I know it seems nit-picky but this is what keeps the framework clean, getting it right the first time instead of getting names wrong or implementation short-sighted, etc. Finally, there is no test for this in moose-test so we might decide to break it next week!

:smile: